server/server.go

```diff
@@ -6,50 +6,185 @@ package server
+	"time"
 
 	"github.com/elastic/elastic-agent-libs/logp"
 	pb "github.com/elastic/elastic-agent-shipper-client/pkg/proto"
 	"github.com/elastic/elastic-agent-shipper-client/pkg/proto/messages"
 	"github.com/elastic/elastic-agent-shipper/queue"
+
+	"github.com/gofrs/uuid"
 )
 
+// Publisher contains all operations required for the shipper server to publish incoming events.
+type Publisher interface {
+	io.Closer
+
+	// AcceptedIndex returns the current sequential index of the accepted events
+	AcceptedIndex() queue.EntryID
+	// PersistedIndex returns the current sequential index of the persisted events
+	PersistedIndex() queue.EntryID
+	// Publish publishes the given event and returns the current accepted index (after this event)
+	Publish(*messages.Event) (queue.EntryID, error)
+}
+
+// ShipperServer contains all the gRPC operations for the shipper endpoints.
+type ShipperServer interface {
+	pb.ProducerServer
+	io.Closer
+}
+
 type shipperServer struct {
-	logger *logp.Logger
+	logger    *logp.Logger
+	publisher Publisher
 
-	queue *queue.Queue
+	uuid string
+
+	close *sync.Once
+	ctx   context.Context
+	stop  func()
 
 	pb.UnimplementedProducerServer
 }
 
-// PublishEvents is the server implementation of the gRPC PublishEvents call
-func (serv shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
-	reply := &messages.PublishReply{}
-	for _, evt := range req.Events {
-		serv.logger.Infof("Got event: %s", evt.String())
-		entryID, err := serv.queue.Publish(evt)
+// NewShipperServer creates a new server instance for handling gRPC endpoints.
+func NewShipperServer(publisher Publisher) (ShipperServer, error) {
+	if publisher == nil {
+		return nil, errors.New("publisher cannot be nil")
+	}
+
+	id, err := uuid.NewV4()
+	if err != nil {
+		return nil, err
+	}
+
+	s := shipperServer{
+		uuid:      id.String(),
+		logger:    logp.NewLogger("shipper-server"),
+		publisher: publisher,
+		close:     &sync.Once{},
+	}
+
+	s.ctx, s.stop = context.WithCancel(context.Background())
+
+	return &s, nil
+}
+
+// GetAcceptedIndex returns the accepted index
+func (serv *shipperServer) GetAcceptedIndex() uint64 {
+	return uint64(serv.publisher.AcceptedIndex())
+}
+
+// GetPersistedIndex returns the persisted index
+func (serv *shipperServer) GetPersistedIndex() uint64 {
+	return uint64(serv.publisher.PersistedIndex())
+}
+
+// PublishEvents is the server implementation of the gRPC PublishEvents call.
+func (serv *shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
+	resp := &messages.PublishReply{
+		Uuid: serv.uuid,
+	}
+
+	// the value in the request is optional
+	if req.Uuid != "" && req.Uuid != serv.uuid {
+		resp.AcceptedIndex = serv.GetAcceptedIndex()
+		resp.PersistedIndex = serv.GetPersistedIndex()
+		serv.logger.Debugf("shipper UUID does not match, all events rejected. Expected = %s, actual = %s", serv.uuid, req.Uuid)
+
+		return resp, nil
```

Comment:
```suggestion
		serv.logger.Debugf("shipper UUID does not match, all events rejected. Expected = %s, actual = %s", serv.uuid, req.Uuid)

		return resp, status.Error(codes.FailedPrecondition, fmt.Errorf("UUID does not match. Expected = %s, actual = %s", serv.uuid, req.Uuid)
```

It still feels like it would be better to return an error to the client here. I'm not sure the returned accepted+persistent indexes have any value if the UUID doesn't match.

- https://pkg.go.dev/google.golang.org/grpc/status#Error
- https://pkg.go.dev/google.golang.org/grpc/codes

Reply:
done in https://github.com/elastic/elastic-agent-shipper/pull/76/commits/4d65122671fec178e72a72204fe6a8d0f28d835e however the code is slightly different.

Reply:
Looks good, thanks!